### DISCUSSION TITLE
Node.js Global Variables are not persisted between getServerSideProps and API Routes

### DISCUSSION THREAD

**Discussion by benjamin-p-newton** [vercel/next.js#44417] [Help] [answered]
Could someone please explain why my server-side node.js global variable is not persisted when called from getServerSideProps and API Routes?

For example, if I construct and export a const ApiClient (w/ a uuid to identify the instance) like so:
```
// apiClient.ts
global['apiclient'] = new ApiClient(uuid())
const apiClient = global['apiclient']
export default apiClient
```

When I initially load a /page and I import/use `apiClient` in `getServerSideProps` the ApiClient is initialised.
However, if I then call a /page/api route and I import/use `apiClient` in the route, a new ApiClient is initialised, rather than using the one that exists on the global variable.

Am I making a fundamental mistake on how I'm initializing the global variable, or is there something else at play here?

**Comment by icyJoseph** [comment, marked as answer]
Well API routes and getServerSideProps, are a special kind of function. Also in general, sharing globals like that is bound to eventually create problems, such as memory leaks, or sensitive data being shared between users. 

Locally, I think you might be able to share globals. As a matter of fact, in any **"traditional" setting where `next start`** is called and a Node process kicks off - you should be able to share globals. 

- https://www.youtube.com/watch?v=sNpX8LLUgHo, this talk highlights one of the issues of shared JS contexts between users.

For example if I do this in GSSP:

```ts
  // @ts-expect-error --
  if (typeof global["foo"] === "undefined") {
    // @ts-expect-error --
    global["foo"] = { foo: "bar" };
  }

```

The on the first GSSP pass, we'll inject foo to the global object. Now if my API route did:

```ts
const seen = {};

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse
) {
  // @ts-expect-error --
  if (typeof seen["foo"] !== "undefined") {
    // @ts-expect-error --
    console.log(seen["foo"] === global["foo"]);
  }
  // @ts-expect-error --
  seen["foo"] = global["foo"];
```

Comparing the global foo seen with the global foo that's still present, we'll print `true` every time after, at least, one request is made to GSSP and then at least one request is made to the API route.

**However**, when you deploy to a service such as Vercel, then GSSP and API routes are handled differently. To keep it simple, they are treated as lambdas. **They run as tasks**, give back their result, stay awake for a couple of seconds, or until timeout, and then they go away, and their entire JS context goes away with them. In this model, from the get go, they cannot even share JS globals, because they run as different tasks. 

- https://www.youtube.com/watch?v=jv4Soz08gBw&t=1334s, the same lambda spins out several times, adjusting to traffic loads

Treating them as serverless lambdas, or tasks, means there's virtually zero runtime required at all times, but when a request arrives, which translates to lower costs, and more easily managed memory.

**Reply by benjamin-p-newton** [reply]
Thank you very much for the clarification.
Funnily enough, I re-read the Next.js documentation about 30 minutes ago and was reminded that all Next.js server-side code deployed to Vercel run as serverless functions - hence no shared global variables!
